### PR TITLE
refactor: `balance_transfer_tx_capacity`

### PR DESCRIPTION
**Pull request by handsome0hell**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Which docs this PR relation**:

Ref #

**Which toolchain this PR adaption**:

No Breaking Change

**Special notes for your reviewer**:

### CODE REVIEW

**Review by EthanYuan — APPROVED** [review]

File: core/rpc/core/src/impl/utils.rs [hunk 1]
```diff
@@ -1371,74 +1325,139 @@ impl<C: CkbRpc> MercuryRpcImpl<C> {
...
+        ctx: &Context,
+        ckb_cells_cache: &mut CkbCellsCache,
+        mut excessed_capacity: u64,
+        transfer_components: &mut TransferComponents,
+    ) -> InnerResult<u64> {
+        if excessed_capacity < MIN_CKB_CAPACITY {
+            loop {
+                if excessed_capacity >= MIN_CKB_CAPACITY {
+                    break;
+                }
```
Comment (quake) [inline core/rpc/core/src/impl/utils.rs:1417]: I don't understand why using `if` and `loop` here?

```
if cond {
    loop {
        if !cond {
            break;
        }
        ...
    }
}
```

```
while cond {
    ...
}
```

is equivalent to

File: core/rpc/core/src/impl/utils.rs [hunk 2]
```diff
@@ -1449,116 +1468,129 @@ impl<C: CkbRpc> MercuryRpcImpl<C> {
...
-                        transfer_components,
-                    )
-                    .await;
-                required_capacity -= capacity_provided as i128;
+        Ok(None)
+    }
+
+    fn find_acp_or_secp_belong_to_items<'a>(
+        &self,
+        cells: &'a mut Vec<packed::CellOutput>,
```
Comment (quake) [inline core/rpc/core/src/impl/utils.rs:1509]: It is very strange to pass a mutable parameter to a function whose method name appears to be readonly (find_xxx), suggest to refactroing this fn params to immutable and return the `index` only:

```
fn find_acp_or_secp_belong_to_items<'a>(
        &self,
        cells: &'a Vec<packed::CellOutput>,
        items: &[Item],
    ) -> Option<usize> {
```

you may use the returned index value to get the mutable element:
```
change_to_existed_cell(&mut transfer_components.outputs[index], change_capacity);
```

Comment (handsome0hell) [reply core/rpc/core/src/impl/utils.rs:1509]: Personally, I think relocating element by index seems like a waste of time (not so elegant?). Anyway, there exists `find` method in rust whose arguments contain `&mut`, see https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find . As a result, I guess maybe returning the element we need is not a bad idea.

**Review by quake — APPROVED** [review]

File: core/rpc/core/src/impl/utils.rs (hunk 2 above)
Comment (quake) [reply core/rpc/core/src/impl/utils.rs:1507]: @EthanYuan I think this PR is ready for merge, I will create a new PR to resolve this issue later.

File: core/rpc/core/src/impl/utils.rs [hunk 3]
```diff
@@ -1371,74 +1325,137 @@ impl<C: CkbRpc> MercuryRpcImpl<C> {
...
+    }
+
+    async fn prepare_capacity_for_new_cell(
+        &self,
+        ctx: &Context,
+        ckb_cells_cache: &mut CkbCellsCache,
+        mut excessed_capacity: u64,
+        transfer_components: &mut TransferComponents,
+    ) -> InnerResult<u64> {
+        if excessed_capacity < MIN_CKB_CAPACITY {
```
Comment (EthanYuan) [inline core/rpc/core/src/impl/utils.rs:1413]:
```suggestion
        if excessed_capacity >= MIN_CKB_CAPACITY {
```

**Pull request merged**